tiledb/storage_format/uri/parse_uri.cc

```diff
@@ -30,113 +30,122 @@
+      : StatusException("ParseURI", message) {
   }
+};
 
-  // Format Version 2: __t1_t2_uuid
-  auto maybe_uuid = name.substr(name.find_last_of('_') + 1);
-  if (maybe_uuid.size() == 32) {
-    return FragmentNameVersion::TWO;
+class FragmentURIException : public StatusException {
+ public:
+  explicit FragmentURIException(const std::string& message)
+      : StatusException("FragmentURI", message) {
   }
+};
 
-  // Else, Format Version 1: __uuid_t1{_t2}
-  return FragmentNameVersion::ONE;
+bool is_element_of(const URI uri, const URI intersecting_uri) {
+  std::string prefix = uri.to_string().substr(
+      0, std::string(uri.c_str()).size() - (uri.last_path_part()).size());
+
+  std::string intersecting_prefix = intersecting_uri.to_string().substr(
+      0,
+      std::string(intersecting_uri.c_str()).size() -
+          (intersecting_uri.last_path_part()).size());
+
+  return (prefix == intersecting_prefix);
 }
 
-Status get_timestamp_range(
-    const URI& uri, std::pair<uint64_t, uint64_t>* timestamp_range) {
-  // Initializations
-  auto name = uri.remove_trailing_slash().last_path_part();
-  *timestamp_range = {0, 0};
+FragmentURI::FragmentURI(const URI& uri)
+    : uri_(uri)
+    , name_(uri.remove_trailing_slash().last_path_part())
+    , timestamp_range_({0, 0})
+    , name_version_(FragmentNameVersion::ONE) {
+  // Ensure input uri is valid (non-empty)
+  if (uri.empty()) {
+    throw FragmentURIException(
+        "Failed to construct FragmentURI; input URI is invalid.");
+  }
 
-  // Cut the suffix
-  auto pos = name.find_last_of('.');
-  name = (pos == std::string::npos) ? name : name.substr(0, pos);
+  // Set name
+  auto pos = name_.find_last_of('.');
+  name_ = (pos == std::string::npos) ? name_ : name_.substr(0, pos);
+  assert(name_.find_last_of('_') != std::string::npos);
+  auto version_str = name_.substr(name_.find_last_of('_') + 1);
 
-  // Get fragment name version
-  auto name_version = get_fragment_name_version(name);
+  // Set name version
+  size_t n = std::count(name_.begin(), name_.end(), '_');
+  if (n == 5) {
+    name_version_ = FragmentNameVersion::THREE;
+  } else if (version_str.size() == 32) {
+    name_version_ = FragmentNameVersion::TWO;
+  }
 
-  // FragmentNameVersion::ONE is equivalent to fragment format version <= 2
-  if (name_version == FragmentNameVersion::ONE) {
-    assert(name.find_last_of('_') != std::string::npos);
-    auto t_str = name.substr(name.find_last_of('_') + 1);
+  // Set fragment version
+  uint32_t ret;
+  std::stringstream ss(version_str);
+  ss >> ret;
+  version_ = ret;
+
+  // Set timestamp range
+  if (name_version_ == FragmentNameVersion::ONE) {
+    version_ = 2;
     sscanf(
-        t_str.c_str(),
+        version_str.c_str(),
         (std::string("%") + std::string(PRId64)).c_str(),
-        (long long int*)&timestamp_range->first);
-    timestamp_range->second = timestamp_range->first;
+        (long long int*)&timestamp_range_.first);
```

Comment:
`sscanf` does not allow any decent error handling.  Probably need to rewrite this.